### ISSUE TITLE
Inference results on Nuscenes dataset

### ISSUE THREAD

**Issue by hailuu684** [Pointcept/PointTransformerV3#144]
Dear Authors,
First of all, thank you for your interesting work. I have learnt a lot from your paper.

I have run your code on the Nuscenes dataset and added the code in test.py to check the inference latency and FPS. However, I got relatively low results in the following table. The average time to predict one frame is approximately 6.7s; for each fragment, it is around 470ms, and GPU memory consumption is around 4.3G. I would like to ask if this result is correct for the Nuscenes dataset. 

In addition, I see the result in Figure 1 in your paper; the inference latency is only 44ms, and memory consumption is 1.2G. May I ask what dataset you ran the code to have those results in?

The below is the modified code to calculate FPS in test.py. Could you please check it if you have time?

Thank you and I am looking forward to your response. 

![Image](https://github.com/user-attachments/assets/0b466cc3-7539-4b4d-8d60-3c6c84470ae9)

`class CustomSemSegTester(TesterBase):

    def test(self):

        assert self.test_loader.batch_size == 1, "Batch size of test data should be 1"
        logger = get_root_logger()
        logger.info(">>>>>>>>>>>>>>>> Start Evaluation >>>>>>>>>>>>>>>>")

        batch_time = AverageMeter()
        intersection_meter = AverageMeter()
        union_meter = AverageMeter()
        target_meter = AverageMeter()

        total_time = 0.0  # Track total inference time
        total_samples = 0  # Track the number of processed samples

        self.model.eval()

        save_path = os.path.join(self.cfg.save_path, "result")
        make_dirs(save_path)
        # create submit folder only on main process
        if (
                self.cfg.data.test.type == "ScanNetDataset"
                or self.cfg.data.test.type == "ScanNet200Dataset"
                or self.cfg.data.test.type == "ScanNetPPDataset"
        ) and comm.is_main_process():
            make_dirs(os.path.join(save_path, "submit"))
        elif (
                self.cfg.data.test.type == "SemanticKITTIDataset" and comm.is_main_process()
        ):
            make_dirs(os.path.join(save_path, "submit"))
        elif self.cfg.data.test.type == "NuScenesDataset" and comm.is_main_process():
            import json

            make_dirs(os.path.join(save_path, "submit", "lidarseg", "test"))
            make_dirs(os.path.join(save_path, "submit", "test"))
            submission = dict(
                meta=dict(
                    use_camera=False,
                    use_lidar=True,
                    use_radar=False,
                    use_map=False,
                    use_external=False,
                )
            )
            with open(
                    os.path.join(save_path, "submit", "test", "submission.json"), "w"
            ) as f:
                json.dump(submission, f, indent=4)

        comm.synchronize()
        record = {}
        # fragment inference
        for idx, data_dict in enumerate(self.test_loader):

            # Track inference start time
            start_time = time.time()

            end = time.time()
            data_dict = data_dict[0]  # current assume batch size is 1
            fragment_list = data_dict.pop("fragment_list")
            segment = data_dict.pop("segment")
            data_name = data_dict.pop("name")
            # pred_save_path = os.path.join(save_path, "{}_pred.npy".format(data_name))
            # if os.path.isfile(pred_save_path):
            #     # logger.info(
            #     #     "{}/{}: {}, loaded pred and label.".format(
            #     #         idx + 1, len(self.test_loader), data_name
            #     #     )
            #     # )
            #     pred = np.load(pred_save_path)
            # else:
            pred = torch.zeros((segment.size, self.cfg.data.num_classes)).cuda()

            for i in range(len(fragment_list)):
                fragment_batch_size = 1
                s_i, e_i = i * fragment_batch_size, min(
                    (i + 1) * fragment_batch_size, len(fragment_list)
                )
                input_dict = collate_fn(fragment_list[s_i:e_i])
                for key in input_dict.keys():
                    if isinstance(input_dict[key], torch.Tensor):
                        input_dict[key] = input_dict[key].cuda(non_blocking=True)
                idx_part = input_dict["index"]
                with torch.no_grad():

                    pred_part = self.model(input_dict)  # (n, k)

                    # pred_part = pred_part["seg_logits"]
                    pred_part = F.softmax(pred_part, -1)
                    if self.cfg.empty_cache:
                        torch.cuda.empty_cache()
                    bs = 0
                    for be in input_dict["offset"]:
                        pred[idx_part[bs:be], :] += pred_part[bs:be]
                        bs = be
            # -----------------------------------------------------------------------
            pred = torch.argmax(pred, dim=1)
            segment = torch.from_numpy(segment).to(pred.device)
            intersection, union, target = intersection_and_union_gpu(
                pred, segment, self.cfg.data.num_classes, self.cfg.data.ignore_index
            )

            intersection_meter.update(intersection)
            union_meter.update(union)
            target_meter.update(target)
            record[data_name] = dict(
                intersection=intersection, union=union, target=target
            )

            mask = union != 0
            iou_class = intersection / (union + 1e-10)

            acc = sum(intersection) / (sum(target) + 1e-10)

            iou = torch.mean(iou_class[mask])  # Directly use PyTorch mean
            m_iou = torch.mean(intersection_meter.sum / (union_meter.sum + 1e-10))
            m_acc = torch.mean(intersection_meter.sum / (target_meter.sum + 1e-10))

            batch_time.update(time.time() - end)

            # Compute FPS

            # Track inference end time
            end_time = time.time()
            total_time += (end_time - start_time)
            total_samples += 1

            # fps = 1 / total_time if total_time > 0 else 0
            fps = total_samples / total_time if total_time > 0 else 0 # spf
            # fps = 1 / spf
            # if idx % 300 == 0:
            #     logger.info(
            #         "Test: {} [{}/{}]-{} "
            #         "Batch {batch_time.val:.3f} ({batch_time.avg:.3f}) "
            #         "Accuracy {acc:.4f} ({m_acc:.4f}) "
            #         "mIoU {iou:.4f} ({m_iou:.4f}) "
            #         "FPS {fps:.4f}".format(
            #             data_name,
            #             idx + 1,
            #             len(self.test_loader),
            #             segment.shape,
            #             batch_time=batch_time,
            #             acc=acc,
            #             m_acc=m_acc,
            #             iou=iou,
            #             m_iou=m_iou,
            #             fps=fps
            #         )
            #     )

            logger.info(
                "Test: {} [{}/{}]-{} "
                "Batch {batch_time.val:.3f} ({batch_time.avg:.3f}) "
                "Accuracy {acc:.4f} ({m_acc:.4f}) "
                "mIoU {iou:.4f} ({m_iou:.4f}) "
                "FPS {fps:.4f}".format(
                    data_name,
                    idx + 1,
                    len(self.test_loader),
                    segment.shape,
                    batch_time=batch_time,
                    acc=acc,
                    m_acc=m_acc,
                    iou=iou,
                    m_iou=m_iou,
                    fps=fps
                )
            )

        logger.info("Syncing ...")
        comm.synchronize()
        record_sync = comm.gather(record, dst=0)

        if comm.is_main_process():
            record = {}
            for _ in range(len(record_sync)):
                r = record_sync.pop()
                record.update(r)
                del r

        # Convert list of tensors to stacked tensor for sum operation
        intersection = torch.sum(torch.stack([meters["intersection"] for _, meters in record.items()]), dim=0)
        union = torch.sum(torch.stack([meters["union"] for _, meters in record.items()]), dim=0)
        target = torch.sum(torch.stack([meters["target"] for _, meters in record.items()]), dim=0)

        # Compute IoU and accuracy using PyTorch operations
        iou_class = intersection / (union + 1e-10)
        accuracy_class = intersection / (target + 1e-10)

        # Compute mean IoU and mean Accuracy using torch.mean()
        mIoU = torch.mean(iou_class)
        mAcc = torch.mean(accuracy_class)

        # Compute overall Accuracy
        allAcc = torch.sum(intersection) / (torch.sum(target) + 1e-10)

        logger.info(
            "Val result: mIoU/mAcc/allAcc {:.4f}/{:.4f}/{:.4f}".format(
                mIoU, mAcc, allAcc
            )
        )

        if not os.path.exists(self.cfg.miou_result_path):
            os.makedirs(self.cfg.miou_result_path)
            print(f"Created directory {self.cfg.miou_result_path}")

        miou_result_file = os.path.join(self.cfg.miou_result_path, "miou_results.txt")
        with open(miou_result_file, "w") as f:
            for i in range(self.cfg.data.num_classes):
                result_line = "Class_{idx} - {name} Result: iou/accuracy {iou:.4f}/{accuracy:.4f}".format(
                    idx=i,
                    name=self.cfg.data.names[i],
                    iou=iou_class[i].item(),  # Convert tensor to float
                    accuracy=accuracy_class[i].item(),  # Convert tensor to float
                )

                # Print to console (optional)
                logger.info(result_line)

                # Write to file
                f.write(result_line + "\n")

        logger.info("<<<<<<<<<<<<<<<<< End Evaluation <<<<<<<<<<<<<<<<<")`

**Comment by WeijianZhang123** [comment]
In my understanding, latency specifically refers to the model inference process, i.e., self.model(input_dict), and excludes data loading and preprocessing.

**Comment by hailuu684** [comment]
> In my understanding, latency specifically refers to the model inference process, i.e., self.model(input_dict), and excludes data loading and preprocessing.

Yeah I understand that. However, I am not sure in test phase, we have segments

**Comment by Gofinge** [comment]
Hey, this is an optimal TTA testing technology. If you are sensitive to testing time. There are many ways to make the process faster. For example: a) Removing all TTA; b) replacing fragment testing with mapping. If you interest in these testing strategy, I can share a more detailed introduction later.

**Comment by hailuu684** [comment]
> Hey, this is an optimal TTA testing technology. If you are sensitive to testing time. There are many ways to make the process faster. For example: a) Removing all TTA; b) replacing fragment testing with mapping. If you interest in these testing strategy, I can share a more detailed introduction later.

Thank you for your reply. I would love to look more into the testing strategy. May I ask what the segments are in the testing phase? Does each fragment mean the point clouds on one scene ( or frame)? I think that because the number of points in each fragment is around 24k, which is the same for one frame or one sweep.

**Comment by Gofinge** [comment]
[Here](https://github.com/Pointcept/Pointcept?tab=readme-ov-file#testing) is a detailed explanation of the testing process. If you are not sensitive to a few thousandth performance gap, I think just mapping the prediction with the "inverse" we record during "GridSampling" should also work.
